File: 0368-largest-divisible-subset/0368-largest-divisible-subset.java
class Solution {
    public List<Integer> largestDivisibleSubset(int[] nums) {
        List<Integer> list = new ArrayList<>();
        Arrays.sort(nums);
        int ans = 0;
        int pos = 0;
        for (int i = 0; i < nums.length; i++) {
            int mx = 1, ind = i;
            for (int j = i - 1; j >= 0; j--) {
                if (nums[i] % nums[j] == 0) {
                    if ((list.get(j) + 1) > mx) {
                        mx = list.get(j) + 1;
                        ind = j;
                    }
                }
            }
            if (mx > ans) {
                ans = mx;
                pos = i;
            }
            list.add(mx);
        }
        List<Integer> res = new ArrayList<>();
        int num = nums[pos];
        for (int i = nums.length - 1; i >= 0; i--) {
            if (num % nums[i] == 0 && list.get(i) == ans) {
                res.add(nums[i]);
                num = nums[i];
                ans--;
            }
        }
        
        return res;
    }
}

File: 0368-largest-divisible-subset/0368-largest-divisible-subset.cpp
class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        int n = nums.size();
        vector<int> prev(n, -1), dp(n, 0);
        int maxSetSize = 0;
        int ansInd = 0;
        dp[0] = 1, prev[0] = -1;
        for (int i = 1; i < nums.size(); i++) {
            int mx = 0;
            int pos = -1;
            for (int j = 0; j < i; j++) {
                if (nums[i] % nums[j] == 0) {
                    if (dp[j] > mx) {
                        mx = dp[j];
                        pos = j;
                    }
                }
            }
            dp[i] = mx + 1;
            prev[i] = pos;
            if (maxSetSize < (mx + 1)) {
                maxSetSize = max(maxSetSize, mx + 1);
                ansInd = i;
            }
        }
        int ind = ansInd;
        vector<int> ans;
        while (ind >= 0) {
            ans.push_back(nums[ind]);
            ind = prev[ind];
        }
        return ans;
    }
};